crypto/x509/x_name.c

```diff
@@ -299,27 +299,28 @@ static int x509_name_ex_print(BIO *out, const ASN1_VALUE **pval,
  * memcmp() of the canonical encoding. By omitting the leading SEQUENCE name
  * constraints of type dirName can also be checked with a simple memcmp().
  */
-
 static int x509_name_canon(X509_NAME *a)
 {
     unsigned char *p;
-    STACK_OF(STACK_OF_X509_NAME_ENTRY) *intname;
+    STACK_OF(STACK_OF_X509_NAME_ENTRY) *intname = NULL;
     STACK_OF(X509_NAME_ENTRY) *entries = NULL;
     X509_NAME_ENTRY *entry, *tmpentry = NULL;
     int i, set = -1, ret = 0, len;
 
     OPENSSL_free(a->canon_enc);
     a->canon_enc = NULL;
-    /* Special case: empty X509_NAME => null encoding */
+
+    /* Special case: empty X509_NAME (NULL-DN) */
     if (sk_X509_NAME_ENTRY_num(a->entries) == 0) {
         a->canon_enclen = 0;
+        if ((a->canon_enc = (unsigned char *)OPENSSL_strdup("")) == NULL)
+            goto oom;
         return 1;
     }
+
     intname = sk_STACK_OF_X509_NAME_ENTRY_new_null();
-    if (intname == NULL) {
-        ERR_raise(ERR_LIB_X509, ERR_R_MALLOC_FAILURE);
-        goto err;
-    }
+    if (intname == NULL)
```

Comment:
This refactoring makes the control flow in the most important case, down at line 365 *much more difficult* to follow. Saving a handful of lines of code versus make the code harder to maintain is the wrong choice. Please don't do this.

Reply:
The whole `ERR_R_MALLOC_FAILURE` business is a tedious and longstanding issue,
as you know - see also, e.g., #5781 and #6251.
I'm tempted to kick out here all `ERR_raise(ERR_LIB_X509, ERR_R_MALLOC_FAILURE)`
calls because the more low-level functions should have taken care of that already.

Reply:
I'm now attempting the notorious malloc error reporting issue once and for all in #14833.

Reply:
I propose not attempting that before 3.0 release.